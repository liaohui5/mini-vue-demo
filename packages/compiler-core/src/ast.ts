import { CREATE_ELEMENT_VNODE } from './runtimeHelpers';

export const enum NodeTypes {
  // 差值表达式
  INTERPOLATION,

  // 简单表达式
  SIMPLE_EXPRESSION,

  // 复合类型表达式
  COMPOUND_EXPRESSION,

  // 元素
  ELEMENT,

  // 文本
  TEXT,

  // 根节点
  ROOT,
}

// 创建根节点
export function createRootASTNode(children: Array<ASTNode>): RootASTNode {
  return {
    type: NodeTypes.ROOT,
    children,
    codegenNode: children[0],
    helpers: [],
  };
}

// 创建插值表达式节点
export function createInterpolationNode(content: Expression): InterpolationASTNode {
  return {
    type: NodeTypes.INTERPOLATION,
    content,
    helpers: [],
  };
}

// 创建复合表达式节点
export function createCompoundExpression(children: Array<ASTNode>): CompoundASTNode {
  return {
    type: NodeTypes.COMPOUND_EXPRESSION,
    children: children,
    helpers: [],
  };
}

// 创建元素表达式节点
export function createElementNode(tag: string): ElementASTNode {
  return {
    type: NodeTypes.ELEMENT,
    tag,
    children: [],
    helpers: [],
  };
}

// 创建文本节点
export function createTextNode(text: string): TextASTNode {
  return {
    type: NodeTypes.TEXT,
    content: text,
    helpers: [],
  };
}

// 创建 元素节点
export function createVNodeCall(ctx: TransformContext, tag: string, props: any, children: Array<any>) {
  ctx.helper(CREATE_ELEMENT_VNODE);
  return {
    type: NodeTypes.ELEMENT,
    helpers: [],
    children,
    tag,
    props,
  };
}
