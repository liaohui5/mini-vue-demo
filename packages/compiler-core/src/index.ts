export * from './compile';
export * from './transform';
export * from './codegen';
