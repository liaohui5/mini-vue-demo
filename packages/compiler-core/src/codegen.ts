import { isString } from '@mini-vue/shared';
import { NodeTypes } from './ast';
import { CREATE_ELEMENT_VNODE, helperNameMapper, TO_DISPLAY_STING } from './runtimeHelpers';

/**
 * 负责生成 render 函数
 */
export function generate(ast: RootASTNode): GenerateResult {
  const ctx = createCodegenContext();
  const fnName = 'render';
  const args = ['_ctx', 'cache'].join(', ');

  // 生成导入代码 const { xxx } = xxx
  genFunctionPreamble(ast, ctx);

  // 生成函数体
  ctx.push(`return function ${fnName}(${args}){return `);
  genNode(ast.codegenNode, ctx);
  ctx.push('}');

  return {
    code: ctx.code,
  };
}

// 生成render函数代码上下文
function createCodegenContext(): CodegenContext {
  return {
    code: '',
    helper(key: symbol): any {
      return helperNameMapper[key];
    },
    push(source: string): void {
      this.code += source;
    },
    newline(): void {
      this.code += '\n';
    },
  };
}

// 根据不同的 node 类型去执行不同的处理犯法
function genNode(node: ASTNode, ctx: CodegenContext): void {
  switch (node.type) {
    case NodeTypes.TEXT:
      genText(<TextASTNode>node, ctx);
      break;
    case NodeTypes.INTERPOLATION:
      genInterpolation(<InterpolationASTNode>node, ctx);
      break;
    case NodeTypes.SIMPLE_EXPRESSION:
      genExpression(<Expression>node, ctx);
      break;
    case NodeTypes.ELEMENT:
      genElement(<ElementASTNode>node, ctx);
      break;
    case NodeTypes.COMPOUND_EXPRESSION:
      genCompoundExpression(<CompoundASTNode>node, ctx);
      break;
    default:
      throw new TypeError(`[codegen]not support ast node type:${node.type}`);
  }
}

// 生成代码导入逻辑
function genFunctionPreamble(ast: ASTNode, ctx: CodegenContext) {
  if (ast.helpers.length > 0) {
    const vueBinging = 'Vue';
    const imports = ast.helpers
      .map((helper: symbol, _index: number, _array: Array<symbol>): string => {
        const helperName = ctx.helper(helper);
        return `${helperName}:_${helperName}`;
      })
      .join(',');

    // 生成 imports 代码: aaa:_aaa, bbb:_bbb
    ctx.push(`const { ${imports} } = ${vueBinging};`);
    ctx.newline();
  }
}

// 文本节点处理
function genText(node: TextASTNode, ctx: CodegenContext): void {
  ctx.push(`'${node.content}'`);
}

// 插值表达式节点处理
function genInterpolation(node: InterpolationASTNode, ctx: CodegenContext): void {
  ctx.push(`_${ctx.helper(TO_DISPLAY_STING)}(`);
  genNode(<ASTNode>node.content, ctx); // -> genExpression -> 将所有逻辑都交给 genNode -> 其他方法只负责生成
  ctx.push(')');
}

// 处理插值表达式的表达式内容
function genExpression(node: Expression, ctx: CodegenContext): void {
  ctx.push(node.content);
}

// 处理元素节点
function genElement(node: ElementASTNode, ctx: CodegenContext): void {
  const { tag, props, children } = node;
  // 最终拼接成一个 createElement('tag', props, children) 的格式;
  ctx.push(`_${ctx.helper(CREATE_ELEMENT_VNODE)}(`);
  const nodes = genNullable([tag, props, children]);
  genNodeList(nodes, ctx);
  ctx.push(')');
}

// 生成
function genNodeList(nodes: Array<ASTNode | string>, ctx: CodegenContext) {
  for (let i = 0, len = nodes.length; i < len; i++) {
    const node = nodes[i];
    if (isString(node)) {
      ctx.push(<string>node); // tag & props
    } else {
      genNode(<ASTNode>node, ctx); // children
    }
    // 最后一个元素不需要加 , createElementVNode('tag', [props], children)
    if (i < len - 1) {
      ctx.push(',');
    }
  }
}

// 如果值为假就为 'null'
function genNullable(args: Array<any>): Array<any> {
  return args.map((arg) => arg || 'null');
}

// 处理复合表达式
function genCompoundExpression(node: CompoundASTNode, ctx: CodegenContext): void {
  for (const child of node.children) {
    if (isString(child)) {
      /* @ts-ignore */
      ctx.push(<string>child);
    } else {
      genNode(child, ctx);
    }
  }
}
