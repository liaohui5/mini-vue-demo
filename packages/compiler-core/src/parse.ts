import { NodeTypes, createInterpolationNode, createElementNode, createTextNode, createRootASTNode } from './ast';

// 标签类型
const enum TagType {
  Start,
  End,
}

// 文本内容中 表达式 开始符号/结束符号 {{ message }}
const startDelimiter: string = '{{';
const closeDelimiter: string = '}}';
const startDelimiterLen: number = startDelimiter.length;
const closeDelimiterLen: number = closeDelimiter.length;

/**
 * 解析 html 为 js object
 */
export function baseParse(content: string): RootASTNode {
  const ctx = createParserContext(content);
  return createRootASTNode(parseChildren(ctx, []));
}

// 创建解析 上下文
function createParserContext(content: string): ParseContext {
  return {
    source: content,
  };
}

// 解析所有子节点
function parseChildren(ctx: ParseContext, ancestors: Array<string>): Array<ASTNode> {
  const nodes: Array<ASTNode> = [];
  let node: ASTNode;
  // let i = 0;
  while (!isEnd(ctx, ancestors)) {
    // i++;
    // if (i > 999) {
    //   console.error('----- infnity loop -----');
    //   break;
    // }

    if (ctx.source.startsWith(startDelimiter)) {
      // 解析 {{message}} 插值表达式
      node = parseInterpolation(ctx);
    } else if (isTag(ctx.source)) {
      // 解析 element <div></div>
      node = parseElement(ctx, ancestors);
    } else {
      // 解析文本  abcd
      node = parseText(ctx);
    }
    nodes.push(node);
  }
  return nodes;
}

// 结束解析子节点循环的条件
function isEnd(ctx: ParseContext, ancestors: Array<string>): boolean {
  // 内容解析完毕才结束
  const s = ctx.source;
  if (s.length === 0) {
    return true;
  }

  // TODO: 为什么此处要遍历所有的 ancestors ?
  for (const item of ancestors) {
    if (startsWithEndTagOpen(ctx, item)) {
      return true;
    }
  }

  return false;
}

// 解析文本节点
function parseText(ctx: ParseContext): TextASTNode {
  // 获取内容 && 向后推进
  // 如果遇到 < {{ 就停止解析 text, 应该去解析表达式和子元素
  let endInex = ctx.source.length;
  const endTokens = [startDelimiter, '<'];
  for (const token of endTokens) {
    const tokenIndex = ctx.source.indexOf(token);
    if (tokenIndex !== -1 && endInex > tokenIndex) {
      // endInex > tokenIndex 取最小值,否则回出错: 如 <div>hi,{{msg}}</div>
      // 如果不取最小值, 那么第二次循环的时候会 hi,{{msg}} 当做一个 text 值
      // 到 < 位置才会停下
      endInex = tokenIndex;
    }
  }
  const text = parseTextData(ctx, endInex);
  return createTextNode(text);
}

// 截取文本并且向后推进
function parseTextData(ctx: ParseContext, len: number): string {
  const text = ctx.source.slice(0, len);
  advanceBy(ctx, len);
  return text;
}

// 判断是否是一个标签
function isTag(str: string): boolean {
  return Boolean(str[0] === '<' && /[a-z]/i.test(str[1]));
}

// 解析 element 元素
function parseElement(ctx: ParseContext, ancestors: Array<string>): ElementASTNode {
  const element = parseTag(ctx, TagType.Start) as ElementASTNode;
  ancestors.push(element.tag);
  element.children = parseChildren(ctx, ancestors); // 递归的去解析
  ancestors.pop();

  // 如果结束标签和当前element不一致, 那么证明没有闭合标签
  if (!startsWithEndTagOpen(ctx, element.tag)) {
    throw new Error(`lost close tag:${element.tag}`);
  }

  parseTag(ctx, TagType.End);
  return element;
}

// 判断当前ctx.source的结束标签是否和 ancestors 中的最后一个是同一个标签(开始和结束)
function startsWithEndTagOpen(ctx: ParseContext, tag: string): boolean {
  return ctx.source.startsWith('</') && ctx.source.slice(2, 2 + tag.length).toLowerCase() === tag.toLowerCase();
}

// 解析 html 标签
function parseTag(ctx: ParseContext, type: TagType): ElementASTNode | void {
  const tagReg = /^<\/?([a-z]+)>/i;

  // matches: [ "<div>", "div", index: 0, input: "<div>..." ]
  const matches = tagReg.exec(ctx.source)!;
  const tag = matches[1];
  advanceBy(ctx, matches[0].length);

  // 如果是闭合标签 </div>
  if (type === TagType.End) {
    return;
  }

  return createElementNode(tag);
}

// 解析插值表达式 {{message}}
function parseInterpolation(ctx: ParseContext): InterpolationASTNode {
  // 截取开始符号 {{message}} -> message}}
  advanceBy(ctx, startDelimiterLen);

  // 截取中间内容 {{message}} -> message
  const rawContent = parseTextData(ctx, ctx.source.indexOf(closeDelimiter));
  const content = rawContent.trim();

  // 截取结束符号 }}
  advanceBy(ctx, closeDelimiterLen);
  return createInterpolationNode({
    type: NodeTypes.SIMPLE_EXPRESSION,
    content,
  });
}

// 向后推进, 已经处理过的就移除
function advanceBy(ctx: ParseContext, len: number) {
  ctx.source = ctx.source.slice(len);
}
