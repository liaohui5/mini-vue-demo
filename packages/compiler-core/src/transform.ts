import { isFunction } from '@mini-vue/shared';
import { NodeTypes } from './ast';
import { TO_DISPLAY_STING } from './runtimeHelpers';

// null: 34

/**
 * transform 主要是遍历 ast 生成 render 函数的函数体内容
 */
export function transform(root: RootASTNode, options: TransformOptions = { nodeTransformers: [] }): void {
  const transformContext = createTransformContext(root, options);
  traverseNode(<ASTNode>root, transformContext);
  createRootCodegen(root);

  // 给当前节点赋值 helpers
  root.helpers = [...transformContext.helpers.keys()];
}

// 给 codegen 创建生成代码节点树
function createRootCodegen(root: RootASTNode): void {
  const child = root.children[0];
  if (NodeTypes.ELEMENT === child.type) {
    root.codegenNode = child.codegenNode;
  } else {
    root.codegenNode = child;
  }
}

// 创建 转换上下文
function createTransformContext(root: ASTNode, options: TransformOptions): TransformContext {
  return {
    root,
    nodeTransformers: options.nodeTransformers || [],
    helpers: new Map(),
    helper(key: symbol): void {
      this.helpers.set(key, 1);
    },
  };
}

// 深度优先, 遍历所有节点(树形结构数据)
function traverseNode(node: ASTNode, ctx: TransformContext): void {
  // 因为 transform 可能会修改节点的结构, 所以应该放到后面执行
  // 所以, 应该收集所有后置执行的函数放到最后执行
  const exitCallbacks: Array<CallableFunction> = [];
  for (const transform of ctx.nodeTransformers) {
    const callback = transform(node, ctx);
    isFunction(callback) && exitCallbacks.push(callback);
  }
  // 在遍历子节点节点之前应该先处理自己的 helpers, 方便让 codegen 生成代码
  switch (node.type) {
    case NodeTypes.TEXT:
      break;
    case NodeTypes.INTERPOLATION:
      // 插值表达式用 toDisplayString 显示字符串
      ctx.helper(TO_DISPLAY_STING);
      break;
    case NodeTypes.ROOT:
    case NodeTypes.ELEMENT:
      // 处理 html 元素节点 或者 root 节点 一定有 children
      traverseChildren(ctx, (<RootASTNode>node).children);
      break;

    default:
      break;
  }

  let i = exitCallbacks.length;
  while (i--) {
    exitCallbacks[i]();
  }
}

// 遍历当前节点的子节点
function traverseChildren(ctx: TransformContext, children: Array<ASTNode>): void {
  for (let i = 0; i < children.length; i++) {
    const childNode = children[i];
    // 有 children 一定是处理 element
    traverseNode(<ElementASTNode>childNode, ctx);
  }
}
