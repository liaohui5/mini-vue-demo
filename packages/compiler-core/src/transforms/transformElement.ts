import { NodeTypes, createVNodeCall } from '../ast';

export function transformElement(node: ASTNode, ctx: TransformContext) {
  if (node.type === NodeTypes.ELEMENT) {
    return () => {
      // ctx.helper(CREATE_ELEMENT_VNODE);
      const { tag, props, children } = <ElementASTNode>node;

      const vnodeTag = `'${tag}'`;

      // TODO: 处理 props
      const vnodeProps = props;

      // 处理 children
      let vnodeChildren = null;
      if (children.length === 1) {
        vnodeChildren = children[0];
      }

      /* @ts-ignore */
      node.codegenNode = createVNodeCall(ctx, vnodeTag, vnodeProps, vnodeChildren);
    };
  }
}
