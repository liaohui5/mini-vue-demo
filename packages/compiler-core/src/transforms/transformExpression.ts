import { NodeTypes } from '../ast';

export function transformExpression(node: InterpolationASTNode) {
  if (node.type === NodeTypes.INTERPOLATION) {
    node.content = processExpression(<Expression>node.content);
  }
}

// 处理表达式内容
function processExpression(expr: Expression): Expression {
  expr.content = `_ctx.${expr.content}`;
  return expr;
}
