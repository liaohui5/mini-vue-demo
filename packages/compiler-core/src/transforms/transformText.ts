import { createCompoundExpression, NodeTypes } from '../ast';
import { isText } from '../utils';

export function transformText(node: ASTNode) {
  // 文本节点才可能包含其他节点类型
  if (node.type !== NodeTypes.ELEMENT) {
    return;
  }

  return () => {
    const children = node.children!;
    let compound: CompoundASTNode | undefined;

    for (let i = 0; i < children.length; i++) {
      const child = children[i];
      if (!isText(child)) {
        continue;
      }

      for (let j = i + 1; j < children.length; j++) {
        // 为了替换当前 text 节点为 compound_express
        const next = children[j];
        if (!isText(next)) {
          // 停止收集
          compound = undefined;
          break;
        }
        if (!compound) {
          // 初始化
          compound = children[i] = createCompoundExpression([child]);
        }

        // 添加其他子节点到 compound
        compound.children.push('+');
        compound.children.push(next);

        // 改变数组长度后, 应该让 j--, 否则可能会越界报错
        children.splice(j, 1);
        j--;
      }
    }
  };
}
