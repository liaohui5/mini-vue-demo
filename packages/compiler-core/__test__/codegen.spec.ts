import { generate } from '../src/codegen';
import { baseParse } from '../src/parse';
import { transformElement } from '../src/transforms/transformElement';
import { transformExpression } from '../src/transforms/transformExpression';
import { transformText } from '../src/transforms/transformText';
import { transform } from '../src/transform';

describe('codegen', () => {
  it('happy path', () => {
    const ast = baseParse('hi');
    transform(ast);
    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });

  it('interpolation', () => {
    const ast = baseParse('{{message}}');

    transform(ast, {
      nodeTransformers: [transformExpression],
    });

    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });

  it('element', () => {
    const ast = baseParse('<div></div>');
    transform(ast, {
      nodeTransformers: [transformElement],
    });
    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });

  it('element + text + interpolation', () => {
    const ast = baseParse('<div>hi,{{message}}</div>');

    transform(ast, {
      nodeTransformers: [transformExpression, transformElement, transformText],
    });

    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });
});
