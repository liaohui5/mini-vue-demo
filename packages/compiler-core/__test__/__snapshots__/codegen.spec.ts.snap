// Vitest Snapshot v1

exports[`codegen > element + text + interpolation 1`] = `
"const { toDisplayString:_toDisplayString,createElementVNode:_createElementVNode } = Vue;
return function render(_ctx, cache){return _createElementVNode('div',null,'hi,'+_toDisplayString(_ctx.message))}"
`;

exports[`codegen > element 1`] = `
"const { createElementVNode:_createElementVNode } = Vue;
return function render(_ctx, cache){return _createElementVNode('div',null,null)}"
`;

exports[`codegen > happy path 1`] = `"return function render(_ctx, cache){return 'hi'}"`;

exports[`codegen > interpolation 1`] = `
"const { toDisplayString:_toDisplayString } = Vue;
return function render(_ctx, cache){return _toDisplayString(_ctx.message)}"
`;
